### DISCUSSION TITLE
Deleting Elements

### DISCUSSION THREAD

**Discussion by RyanDaley** [GeometryGym/GeometryGymIFC#76] [Q&A]
Is it possible to delete IfcEntities?  The examples show clearly how to make new entities, but it isn't clear how to remove/delete unwanted elements.  The best I can come up with is removing the references to the IfcEntity.

In the following example, I am trying to delete the old IfcBuildingStorey, and make a new one with the same RelatedElements but with different properties, name, description, ect.  The only way I could think to delete the old IfcBuildingStorey was to remove all the RelatedElements  from the buildingstorey, but also remove the reference to the IfcBuildingStorey from the IfcBuilding. 

Is there a better way to do this?

```
BuildingStoreyConcept myBuildingStoreyConcept = BuildingStoreyCollection.FirstOrDefault(x => x.Name == "Floor 1");
var productListForRebuild = myBuildingStoreyConcept.RelContainedInSpatialStructure.RelatedElements.ToList();

IfcBuildingStorey oldBuildingStorey = db.OfType<IfcBuildingStorey>().Where(x=>x.Name == "Floor 1").FirstOrDefault();
IfcRelContainedInSpatialStructure relContains = oldBuildingStorey.ContainsElements.First();
List<IfcProduct> productList = relContains.RelatedElements.ToList();
foreach(IfcProduct ifcProduct in productList)
{
    relContains.RelatedElements.Remove(ifcProduct);
}

IfcBuildingStorey newBuildingStorey = new IfcBuildingStorey(myBuildingStoreyConcept.RelatingBuilding, myBuildingStoreyConcept.Name, myBuildingStoreyConcept.Elevation);
//BUILD NEW IFCRELDEFINESBYPROPERTIES
foreach (IfcRelDefinesByProperties relDefinesByProperties in myBuildingStoreyConcept.RelDefinedByProperties)
{
    IfcRelDefinesByProperties newRelDefinesByProperties = new IfcRelDefinesByProperties(relDefinesByProperties.RelatedObjects.First(), relDefinesByProperties.RelatingPropertyDefinition.First());
}
//NEW REL AGGREGATES OBJECTS TO BUILDING STORIES
IfcRelAggregates relAgg2 = new IfcRelAggregates(newBuildingStorey, productListForRebuild); 
```

**Comment by jmirtsch** [comment]
I'd recommend a couple of slightly different strategies, let's see if they are adequate.

First, rather than delete the building storey, it would be best to apply the modifications to the existing.
You can rename, change description etc.  But perhaps your requirements are that this isn't the right strategy.
Strictly speaking product placement is required to be a traversal of the spatial structure (ie a product placement is relative to say the space or storey that contains it).  Most viewers don't rely on this assumption, but you might want to bear it in mind if you do delete and create anew, you might want to retain the previous story placement.  A validation will flag placements that aren't related to a product.

I do have code in the closed source part of the Geometry Gym developments to delete entities.  There are a variety of
inverse relationships (such as material assignment, systems etc, particularly for elements).  It's not a trivial method, especially when aspects such as geometry might be common to multiple elements.  At present I haven't included it in the open source toolkit.  
However, you can remove the elements from the spatial structure as you show, and then you can "duplicate" the project to ensure
only active data is included in the saved file when writing (without lots of orphaned concepts that were only assigned to obsolete entities).

Just be careful also about your relationships.  Elements are contained in a spatial structure, not an aggregation.  I find the KIT model checker a good way to review the model to ensure the right relationships are used. 
ie IfcRelAggregates relAgg2 = new IfcRelAggregates(newBuildingStorey, productListForRebuild);
should be new IfcRelContainedInSpatialStructure to relate products to the storey.

**Reply by RyanDaley** [reply]
Thanks @jmirtsch for the thorough answer.  

The desire to remove and replace comes from complicated data relationships on my end.  I am trying to change which IfcSpatialElement many of the IfcEntities are associated with.  For instance, I often times want to move IfcEntities from one IfcBuildingStorey to another IfcBuildingStorey, and at the same time, change which IfcBuilding the IfcBuildingStorey is associated with.  

With changing relationships of this complexity, it often times is arduous to keep track of the old reference, so I can remove it, and then creating the new IfcRelAggregates.  It seemed like it would be easier just to delete all of the IfcSpatialElements, and rebuild all of the relationships exactly how I want them from scratch.

**Comment by jmirtsch** [comment]
I figured it was probably something like that.
One method I didn't mention, was the AddElement method which can be called on the new storey. 
 https://github.com/GeometryGym/GeometryGymIFC/blob/a45248f153c6b43d077b1a6174a29c11312cf7c9/Core/IFC/IFC%20P.cs#L1238
It will remove any existing containment or aggregation relationship.

For the project duplication, I should make the method public and open source.  But the code to duplciate the IfcProject looks like this:

	DatabaseIfc db = new DatabaseIfc(mDatabase);
	db.Factory.Duplicate(ifcProject, options);

**Reply by RyanDaley** [reply]
> I figured it was probably something like that. One method I didn't mention, was the AddElement method which can be called on the new storey.
> 
> https://github.com/GeometryGym/GeometryGymIFC/blob/a45248f153c6b43d077b1a6174a29c11312cf7c9/Core/IFC/IFC%20P.cs#L1238
> 
> 
> It will remove any existing containment or aggregation relationship.

> ```

Okay, so if I use the AddElement method, it will add the IfcEntity to the new IfcBuildingStorey, and simultaneously remove the reference to the old IfcBuildingStorey (that still exists in the project).

It looks like IfcBuildingStorey, IfcBuilding, and IfcSite all have an .AddElement() method.  However, IfcProject does not.  I was hoping to add more than one IfcSite, but maybe this isn't possible within the IfcSchema anyway.

**Reply by jmirtsch** [reply]
There are implementers agreements in IFC2x3 that a project should only have one site, but the schema permits multiple sites and I believe
most viewers will work ok with multiple.

You are right that IfcProject doesn't have this method, but you can use the aggregation for the entity.  Yes, I think the AddElement provides the basis of what you need otherwise.

**Reply by RyanDaley** [reply]
When adding new floors to my IfcBuilding using the IfcBuilding.AddElement() method, it doesn't create an IfcRelAggregates, it only creates and IfcRelContainedInSpatialStructure. 

![image](https://user-images.githubusercontent.com/65096230/194886274-186bd2ba-3a1e-474f-a2a0-14346f68d65c.png)

This seems like an issue, because when I try to Get the IFcBuildingStorey later, it is missing the .Decomposes property. 

![image](https://user-images.githubusercontent.com/65096230/194886956-d203d3d5-eaa9-4df5-905f-1a56ef1ccb8c.png)

I solve this by manually creating a new IfcRelAggregates(IfcBuilding, IfcBuildingStorey) after I do the IfcBuilding.AddElement().  I guess this is still better because the AddElement() removes the old IfcRelAggregates

**Reply by jmirtsch** [reply]
Yes, this isn't utilizing the right relationship in this situation.  I've just pushed a commit that should work better, if you need a nuget package built, let me know.

Thanks,

Jon

**Reply by RyanDaley** [reply]
An updated Nuget package would be appreciated.

**Reply by jmirtsch** [reply]
I've uploaded a new build of the nuget package.  Let me know you spot anything else amiss.

**Reply by RyanDaley** [reply]
@jmirtsch , I've been working with trying to move IfcEntity around to different IfcBuildingStorey.  It seems like using the .AddElement() method in conjunction with creating a new IfcRelAggregates works for most IfcEntity.  However, it does not work for IfcElementAssembly.

It would seem, moving IfcElementAssembly to a new IfcBuildingStorey requires also making a new IfcRelContainedInSpatialStructure.  I might have to also remove reference to the old IfcRelContainedInSpatialStructure (I need to test this relationship more).  

```
builtElement.AddElement(childProd);
IfcRelAggregates relAgg2 = new IfcRelAggregates(builtElement, childProd);
IfcRelContainedInSpatialStructure relCont = new IfcRelContainedInSpatialStructure(childProd, builtElement as IfcSpatialElement); 
```

The above code works for all IfcElementAssembly except the ones that belong to an 'Undefined' IfcBuildingStorey.  Then, it creates two IfcRelContainedInSpatialStructure for the same object.  Very strange behavior.  Maybe, the AddElement() method does not properly remove the old references for these objects, but this also doesn't make sense because they belong to an actual IfcBuildingStorey called 'Undefined'

![image](https://user-images.githubusercontent.com/65096230/205594082-a54d0613-2a11-4830-ac76-55757b9dccdc.png)
